File: apps/rewarded-keeper/src/app/types/publisher.ts
import { Timestamp } from "firebase/firestore";
import { Month } from "./month";
import { Report } from "./report";

export enum PublisherActivityStatus  {
  Active,
  Irregular,
  Inactive,
}

export interface Publisher {
  id?: string;
  name: string;
  firstName: string;
  lastName: string;
  groupId: string | 'unafiliated';
  address: string;
  telephone: string;
  emergencyPhone: string;
  emailAddress: string;
  isElder?: boolean;
  isMinisterialServant?: boolean;
  isSpecialServant?: boolean;
  isRegularPioneer?: boolean;
  auxilaryPionierFor?: string[];
  isPermanentAuxilaryPioneer?: boolean;
  activityStatus: PublisherActivityStatus;
  birthDate?: Timestamp;
  baptismDate?: Timestamp;
  reports?: Report[]; // New field for migrating reports from separate collection to publisher arrays
  congregationId?: number; // Multi-tenancy: congregation this publisher belongs to (stores congregation.congregationNumber)
}

/**
 * Determines whether a publisher is auxilary pionier for the given month.
 *
 * @param {Publisher|undefined} publisher - The publisher object to check.
 * @param {string} monthId - The ID of the month to check.
 * @returns {boolean} - Boolean value indicating whether the publisher is auxilary pionier for the given month.
 */
export const isPublisherAuxilaryPionierForMonth = (
  publisher: Publisher | undefined,
  monthId: string
) => {
  return publisher?.isPermanentAuxilaryPioneer || publisher?.auxilaryPionierFor?.includes(monthId) || false;
};


/**
 * Checks if a publisher is a special publisher for a given month.
 *
 * @param {Publisher} publisher - The publisher to check.
 * @param {Month} [month] - The month to check for. If not provided, defaults to current month.
 * @returns {boolean} - True if the publisher is a special publisher for the given month, otherwise false.
 */
export const isSpecialPublisher = (publisher?: Publisher, month?: Month) => {
  return isPublisherAuxilaryPionierForMonth(publisher, month?.getKey() || '') ||
    publisher?.isPermanentAuxilaryPioneer ||
    publisher?.isRegularPioneer ||
    publisher?.isSpecialServant || false;
}

File: server/entities/publisher.go
package entities

import "time"

type Report struct {
  id            string
  publisherId   string
  monthId       string
  publications  int
  active        bool /**Use if the publisher has preached during the month. @since November 1st, 2023 */
  videos        int
  hours         int
  visits        int
  courses       int
  comment       string
  date          time.Time
  submitted     bool
  isFirstReport bool
  isAPReport    bool
}

type Publisher struct {
  Id                   string   `json:"id" bson:"_id,omitempty"`
  Name                 string   `json:"name" bson:"name,omitempty"`
  FirstName            string   `json:"firstName" bson:"firstName,omitempty"`
  LastName             string   `json:"lastName" bson:"lastName,omitempty"`
  GroupId              string   `json:"groupId" bson:"groupId,omitempty"`
  Address              string   `json:"address" bson:"address,omitempty"`
  Telephone            string   `json:"telephone" bson:"telephone,omitempty"`
  EmergencyPhone       string   `json:"emergencyPhone" bson:"emergencyPhone,omitempty"`
  EmailAddress         string   `json:"emailAddress" bson:"emailAddress,omitempty"`
  IsElder              bool     `json:"isElder" bson:"isElder,omitempty"`
  IsMinisterialServant bool     `json:"isMinisterialServant" bson:"isMinisterialServant,omitempty"`
  IsSpecialServant     bool     `json:"isSpecialServant" bson:"isSpecialServant,omitempty"`
  IsRegularPioneer     bool     `json:"isRegularPioneer" bson:"isRegularPioneer,omitempty"`
  ApMonths             []string `json:"apMonths" bson:"apMonths,omitempty"`
  IsPermanentAP        bool     `json:"isPermanentAP" bson:"isPermanentAP,omitempty"`
  ActivityStatus       int      `json:"activityStatus" bson:"activityStatus,omitempty"`
  Reports              []Report `json:"reports" bson:"reports,omitempty"`
  RealmId              string   `json:"realmId" bson:"realmId,omitempty"`
}
